kusto/kcsb.go

```diff
@@ -11,191 +10,351 @@ import (
+	AZCLI                            bool   `json:"AZCLI"`
+	MSIAuthentication                bool   `json:"MSIAuthentication"`
+	ManagedServiceIdentity           string `json:"ManagedServiceIdentity"`
+	InteractiveLogin                 bool   `json:"InteractiveLogin"`
+	RedirectURL                      string `json:"RedirectURL"`
+	ClientOptions                    *azcore.ClientOptions
 }
 
-// environmental variables
-const (
-	tenantIdEnvVariable     = "AZURE_TENANT_ID"
-	clientIdEnvVariable     = "AZURE_CLIENT_ID"
-	clientSecretEnvVariable = "AZURE_CLIENT_SECRET"
-)
-
 // params mapping
 const (
-	tenantIDStr      string = "TenantID"
-	clientIDStr      string = "ClientID"
-	clientSecretStr  string = "ClientSecret"
-	appCertStr       string = "ApplicationCertificate"
-	appCertKeyStr    string = "ApplicationCertificateKey"
-	usernameStr      string = "Username"
-	userPasswordStr  string = "UserPassword"
-	appTokenStr      string = "ApplicationToken"
-	clientOptionsStr string = "ClientOptions"
-	sendCertChainStr string = "SendCertificateChain"
-	managedIDStr     string = "ManagedIdentityID"
+	dataSource                       string = "DataSource"
+	aadUserId                        string = "AADUserID"
+	password                         string = "Password"
+	applicationClientId              string = "ApplicationClientId"
+	applicationKey                   string = "ApplicationKey"
+	applicationCertificate           string = "ApplicationCertificate"
+	applicationCertificateThumbprint string = "ApplicationCertificateThumbprint"
+	sendCertificateChain             string = "SendCertificateChain"
+	authorityId                      string = "AuthorityId"
+	applicationToken                 string = "ApplicationToken"
+	userToken                        string = "UserToken"
+	msiAuth                          string = "MSIAuthentication"
+	managedServiceIdentity           string = "ManagedServiceIdentity"
+	azCli                            string = "AZCLI"
+	interactiveLogin                 string = "InteractiveLogin"
+	domainHint                       string = "RedirectURL"
 )
 
-// authtype mapping
-const (
-	envAuth        string = "EnvironmentVars"
-	azCliAuth      string = "AzCLI"
-	managedIDAuth  string = "ManagedIdentity"
-	clientCredAuth string = "ClientCredentials"
-	appCertAuth    string = "ApplicationCetrifiate"
-	unamePassAuth  string = "UsernamePassword"
-	appTokenAuth   string = "ApplicationToken"
-)
+func assertIfEmpty(key string, value string) {
+	if isEmpty(value) {
+		panic(fmt.Sprintf("Error: %s cannot be null", key))
+	}
+}
+
+func contains(list []string, tofind string) bool {
+	for _, s := range list {
+		if tofind == s {
+			return true
+		}
+	}
+	return false
+}
+
+func assignValue(kcsb *connectionStringBuilder, rawKey string, value string) {
+	rawKey = strings.ToLower(strings.Trim(rawKey, " "))
+	if contains([]string{"datasource", "data source", "addr", "address", "network address", "server"}, rawKey) {
```

Comment:
Can this be simplified ? Essentially we have a KV string that is parsed to a JSON ?
https://stackoverflow.com/questions/35663892/parse-string-into-map-golang

Reply:
There are multiple keys that can fit each item. I think using a map is correct here.